File: test/modules/usermanagement/domain/organization_domain_test.dart
import 'package:arptc_connect/modules/usermanagement/domain/organization.dart';
import 'package:arptc_connect/modules/usermanagement/domain/organization_assignment.dart';
import 'package:arptc_connect/modules/usermanagement/domain/organization_query.dart';
import 'package:arptc_connect/modules/usermanagement/domain/organization_unit.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('assignment query identity includes organization and subject', () {
    const first = OrganizationAssignmentQuery.agent(
      organizationId: 'org-a',
      agentId: 'agent-a',
    );
    const same = OrganizationAssignmentQuery.agent(
      organizationId: 'org-a',
      agentId: 'agent-a',
    );
    const otherOrganization = OrganizationAssignmentQuery.agent(
      organizationId: 'org-b',
      agentId: 'agent-a',
    );
    const unit = OrganizationAssignmentQuery.unit(
      organizationId: 'org-a',
      unitId: 'agent-a',
    );

    expect(first, same);
    expect(first.hashCode, same.hashCode);
    expect(first, isNot(otherOrganization));
    expect(first, isNot(unit));
    expect(first.isValid, isTrue);
  });

  group('OrganizationHierarchyPolicy', () {
    const policy = OrganizationHierarchyPolicy();

    test('accepts the default Department, Service, Bureau hierarchy', () {
      expect(
        policy.acceptsParent(
          childType: OrganizationUnitType.department,
          parentType: null,
        ),
        isTrue,
      );
      expect(
        policy.acceptsParent(
          childType: OrganizationUnitType.service,
          parentType: OrganizationUnitType.department,
        ),
        isTrue,
      );
      expect(
        policy.acceptsParent(
          childType: OrganizationUnitType.bureau,
          parentType: OrganizationUnitType.service,
        ),
        isTrue,
      );
    });

    test('rejects invalid and skipped hierarchy levels', () {
      expect(
        policy.acceptsParent(
          childType: OrganizationUnitType.bureau,
          parentType: OrganizationUnitType.department,
        ),
        isFalse,
      );
      expect(
        policy.acceptsParent(
          childType: OrganizationUnitType.service,
          parentType: null,
        ),
        isFalse,
      );
      expect(
        policy.allowedChildTypes(OrganizationUnitType.bureau),
        isEmpty,
      );
    });

    test('declares the only next level for each default hierarchy node', () {
      expect(
        policy.allowedChildTypes(null),
        [OrganizationUnitType.department],
      );
      expect(
        policy.allowedChildTypes(OrganizationUnitType.department),
        [OrganizationUnitType.service],
      );
      expect(
        policy.allowedChildTypes(OrganizationUnitType.service),
        [OrganizationUnitType.bureau],
      );
      expect(
        policy.allowedChildTypes(OrganizationUnitType.bureau),
        isEmpty,
      );
    });

    test('reports required parent types used by hierarchy dialogs', () {
      expect(
        policy.requiredParentType(OrganizationUnitType.department),
        isNull,
      );
      expect(
        policy.requiredParentType(OrganizationUnitType.service),
        OrganizationUnitType.department,
      );
      expect(
        policy.requiredParentType(OrganizationUnitType.bureau),
        OrganizationUnitType.service,
      );
    });

    test('custom units require a parent and may form custom chains', () {
      expect(
        policy.acceptsParent(
          childType: OrganizationUnitType.custom,
          parentType: null,
        ),
        isFalse,
      );
      expect(
        policy.acceptsParent(
          childType: OrganizationUnitType.custom,
          parentType: OrganizationUnitType.custom,
        ),
        isTrue,
      );
      expect(
        policy.allowedChildTypes(OrganizationUnitType.custom),
        [OrganizationUnitType.custom],
      );
    });
  });

  test('Organization reads status and schema fields', () {
    final organization = Organization.fromMap(
      {
        'code': 'ARPTC',
        'name': 'ARPTC',
        'description': 'Regulatory authority',
        'status': 'inactive',
        'schemaVersion': 2,
      },
      id: 'org-arptc',
    );

    expect(organization.id, 'org-arptc');
    expect(organization.status, OrganizationStatus.inactive);
    expect(organization.nameLower, 'arptc');
    expect(organization.schemaVersion, 2);
  });

  test('OrganizationUnit reads authoritative path projections', () {
    final unit = OrganizationUnit.fromMap(
      {
        'organizationId': 'org-arptc',
        'type': 'BUREAU',
        'code': 'B-SUP',
        'name': 'Support Desk',
        'parentUnitId': 'service-it',
        'parentUnitType': 'SERVICE',
        'ancestorUnitIds': ['department-it', 'service-it'],
        'pathUnitIds': ['department-it', 'service-it', 'bureau-support'],
        'pathNames': ['IT', 'Operations', 'Support Desk'],
        'scopeKeys': [
          'org:org-arptc',
          'unit:department-it',
          'unit:service-it',
          'unit:bureau-support',
        ],
        'depth': 2,
        'status': 'ACTIVE',
      },
      id: 'bureau-support',
    );

    expect(unit.type, OrganizationUnitType.bureau);
    expect(unit.parentUnitType, OrganizationUnitType.service);
    expect(unit.breadcrumb, 'IT / Operations / Support Desk');
    expect(unit.depth, 2);
    expect(unit.scopeKeys, contains('unit:department-it'));
    expect(() => unit.pathUnitIds.add('forged'), throwsUnsupportedError);
  });

  test('OrganizationUnit reads permanent and acting leadership separately', () {
    final unit = OrganizationUnit.fromMap(
      {
        'organizationId': 'org-arptc',
        'type': 'SERVICE',
        'code': 'S-OPS',
        'name': 'Operations',
        'headUserId': 'uid-permanent',
        'headAssignmentId': 'head-permanent',
        'actingHeadUserId': 'uid-acting',
        'actingHeadAssignmentId': 'head-acting',
        'actingHeadEndsAt': '2026-08-31T00:00:00.000Z',
        'status': 'ACTIVE',
      },
      id: 'service-operations',
    );

    expect(unit.headUserId, 'uid-permanent');
    expect(unit.headAssignmentId, 'head-permanent');
    expect(unit.actingHeadUserId, 'uid-acting');
    expect(unit.actingHeadAssignmentId, 'head-acting');
    expect(unit.actingHeadEndsAt, DateTime.utc(2026, 8, 31));
  });

  test('OrganizationAssignment keeps historical snapshots', () {
    final assignment = OrganizationAssignment.fromMap(
      {
        'organizationId': 'org-arptc',
        'agentId': 'uid-1',
        'unitId': 'bureau-support',
        'unitType': 'BUREAU',
        'unitName': 'Support Desk',
        'ancestorUnitIds': ['department-it', 'service-it'],
        'pathUnitIds': ['department-it', 'service-it', 'bureau-support'],
        'pathNames': ['IT', 'Operations', 'Support Desk'],
        'scopeKeys': ['org:org-arptc', 'unit:bureau-support'],
        'assignmentType': 'HEAD',
        'isPrimary': false,
        'isActing': true,
        'status': 'ACTIVE',
        'startsAt': '2026-08-01T00:00:00.000Z',
        'reason': 'Acting appointment',
      },
      id: 'assignment-1',
    );

    expect(assignment.assignmentType, OrganizationAssignmentType.head);
    expect(assignment.isActing, isTrue);
    expect(assignment.isCurrent, isTrue);
    expect(assignment.pathNames.last, 'Support Desk');
  });

  test('ended assignments remain historical and are not current', () {
    final assignment = OrganizationAssignment.fromMap(
      {
        'organizationId': 'org-arptc',
        'agentId': 'uid-1',
        'unitId': 'bureau-support',
        'unitType': 'BUREAU',
        'unitName': 'Support Desk',
        'pathNames': ['IT', 'Operations', 'Support Desk'],
        'assignmentType': 'MEMBER',
        'isPrimary': true,
        'status': 'ENDED',
        'startsAt': '2025-01-01T00:00:00.000Z',
        'endsAt': '2026-07-31T00:00:00.000Z',
        'endedAt': '2026-07-31T00:00:00.000Z',
        'endedBy': 'manager-1',
        'reason': 'Transfer to Infrastructure',
      },
      id: 'assignment-old',
    );

    expect(assignment.isCurrent, isFalse);
    expect(assignment.isPrimary, isTrue);
    expect(assignment.endsAt, DateTime.utc(2026, 7, 31));
    expect(assignment.endedBy, 'manager-1');
    expect(assignment.pathNames, ['IT', 'Operations', 'Support Desk']);
    expect(() => assignment.pathNames.add('Changed'), throwsUnsupportedError);
  });

  test('OrganizationPageRequest rejects unbounded page sizes', () {
    expect(
      () => OrganizationPageRequest(
        limit: OrganizationPageRequest.maximumLimit + 1,
      ),
      throwsAssertionError,
    );
    expect(
      const OrganizationPageRequest(search: '  Support ').normalizedSearch,
      'support',
    );
  });

  test('OrganizationPageRequest accepts the documented maximum page size', () {
    const request = OrganizationPageRequest(
      limit: OrganizationPageRequest.maximumLimit,
      afterNameLower: 'support',
      afterId: 'unit-50',
    );

    expect(request.limit, 100);
    expect(request.afterNameLower, 'support');
    expect(request.afterId, 'unit-50');
    expect(() => OrganizationPageRequest(limit: 0), throwsAssertionError);
  });

  test('OrganizationPage exposes stable cursor availability', () {
    const cursor = OrganizationPageCursor(
      nameLower: 'support',
      id: 'unit-50',
    );
    const populated = OrganizationPage<String>(
      items: ['unit-1'],
      nextCursor: cursor,
    );
    const terminal = OrganizationPage<String>(
      items: ['unit-2'],
      nextCursor: null,
    );

    expect(populated.hasMore, isTrue);
    expect(populated.nextCursor?.id, 'unit-50');
    expect(terminal.hasMore, isFalse);
  });
}

File: functions/test/organization_domain.test.js
'use strict';

const assert = require('node:assert/strict');
const test = require('node:test');

const {
  buildAgentAuthorizationProjection,
  buildAgentDirectoryProjection,
  buildAgentOrganizationProjection,
  buildAssignmentDocument,
  buildOrganizationUnitProjection,
  normalizeCode,
  validateOrganizationInput,
} = require('../src/organization_domain');

const organization = {
  id: 'org-arptc',
  name: 'ARPTC',
  status: 'ACTIVE',
};

test('normalizes and validates organization metadata', () => {
  assert.equal(normalizeCode(' arptc connect '), 'ARPTC-CONNECT');
  assert.deepEqual(validateOrganizationInput({
    code: ' arptc ',
    name: ' ARPTC ',
    description: ' Authority ',
  }), {
    code: 'ARPTC',
    name: 'ARPTC',
    nameLower: 'arptc',
    description: 'Authority',
  });
});

test('builds a server-owned Department -> Service -> Bureau path', () => {
  const department = withId('department-it', buildOrganizationUnitProjection({
    unitId: 'department-it',
    organization,
    input: {
      organizationId: organization.id,
      type: 'DEPARTMENT',
      code: 'D-IT',
      name: 'Information Technology',
    },
    parent: null,
  }));
  const service = withId('service-operations', buildOrganizationUnitProjection({
    unitId: 'service-operations',
    organization,
    input: {
      organizationId: organization.id,
      type: 'SERVICE',
      code: 'S-OPS',
      name: 'IT Operations',
      parentUnitId: department.id,
    },
    parent: department,
  }));
  const bureau = withId('bureau-helpdesk', buildOrganizationUnitProjection({
    unitId: 'bureau-helpdesk',
    organization,
    input: {
      organizationId: organization.id,
      type: 'BUREAU',
      code: 'B-HD',
      name: 'Help Desk',
      parentUnitId: service.id,
    },
    parent: service,
  }));

  assert.deepEqual(bureau.ancestorUnitIds, [department.id, service.id]);
  assert.deepEqual(bureau.pathUnitIds, [department.id, service.id, bureau.id]);
  assert.deepEqual(bureau.pathNames, [
    'Information Technology',
    'IT Operations',
    'Help Desk',
  ]);
  assert.equal(bureau.depth, 2);
  assert.deepEqual(bureau.scopeKeys, [
    'org:org-arptc',
    'unit:department-it',
    'unit:service-operations',
    'unit:bureau-helpdesk',
  ]);
});

test('rejects invalid, skipped, and cross-organization parents', () => {
  const department = {
    id: 'department-it',
    organizationId: organization.id,
    type: 'DEPARTMENT',
    status: 'ACTIVE',
    pathUnitIds: ['department-it'],
    pathNames: ['IT'],
  };
  assert.throws(
    () => buildOrganizationUnitProjection({
      unitId: 'bureau-helpdesk',
      organization,
      input: {
        organizationId: organization.id,
        type: 'BUREAU',
        code: 'B-HD',
        name: 'Help Desk',
        parentUnitId: department.id,
      },
      parent: department,
    }),
    (error) => error.code === 'failed-precondition' &&
      error.message.includes('service'),
  );
  assert.throws(
    () => buildOrganizationUnitProjection({
      unitId: 'service-it',
      organization,
      input: {
        organizationId: organization.id,
        type: 'SERVICE',
        code: 'S-IT',
        name: 'IT',
        parentUnitId: department.id,
      },
      parent: { ...department, organizationId: 'org-other' },
    }),
    (error) => error.code === 'failed-precondition' &&
      error.message.includes('different organization'),
  );
});

test('builds an immutable assignment snapshot and agent projections', () => {
  const pathUnits = [
    activeUnit('department-it', 'DEPARTMENT', 'Information Technology', [
      'department-it',
    ]),
    activeUnit('service-operations', 'SERVICE', 'IT Operations', [
      'department-it',
      'service-operations',
    ]),
    activeUnit('bureau-helpdesk', 'BUREAU', 'Help Desk', [
      'department-it',
      'service-operations',
      'bureau-helpdesk',
    ]),
  ];
  const unit = {
    ...pathUnits.at(-1),
    ancestorUnitIds: ['department-it', 'service-operations'],
    pathNames: pathUnits.map((entry) => entry.name),
    scopeKeys: [
      'org:org-arptc',
      'unit:department-it',
      'unit:service-operations',
      'unit:bureau-helpdesk',
    ],
  };
  const startsAt = new Date('2026-08-11T00:00:00.000Z');
  const assignment = buildAssignmentDocument({
    assignmentId: 'assignment-1',
    agentId: 'uid-1',
    unit,
    startsAt,
    reason: 'Initial placement',
    actorUid: 'manager-1',
  });
  const placement = buildAgentOrganizationProjection({
    organization,
    unit,
    pathUnits,
    assignmentId: 'assignment-1',
  });
  const directory = buildAgentDirectoryProjection({
    agentId: 'uid-1',
    agent: {
      firstName: 'Aline',
      name: 'Mbuyi',
      postName: 'Kanku',
      email: 'ALINE@ARPTC.CD',
      jobTitle: 'Support analyst',
      isActive: true,
    },
    placement,
  });
  const authorization = buildAgentAuthorizationProjection({
    agent: {
      isActive: true,
      modulePermissions: {
        ticketing: 'MANAGER',
        news: 'NONE',
      },
    },
    placement,
  });

  assert.equal(assignment.status, 'ACTIVE');
  assert.equal(assignment.unitName, 'Help Desk');
  assert.equal(placement.departmentId, 'department-it');
  assert.equal(placement.serviceId, 'service-operations');
  assert.equal(placement.bureauId, 'bureau-helpdesk');
  assert.equal(directory.displayName, 'Aline Mbuyi Kanku');
  assert.equal(directory.email, 'aline@arptc.cd');
  assert.equal(directory.primaryOrganizationUnitId, 'bureau-helpdesk');
  assert.equal(directory.matricule, undefined);
  assert.equal(directory.modulePermissions, undefined);
  assert.deepEqual(authorization.moduleRoleKeys, ['ticketing:MANAGER']);
  assert.ok(authorization.scopeRoleKeys.includes(
    'unit:department-it|ticketing:MANAGER',
  ));
});

test('requires an end date for acting leadership', () => {
  const unit = {
    ...activeUnit('bureau-helpdesk', 'BUREAU', 'Help Desk', [
      'department-it',
      'service-operations',
      'bureau-helpdesk',
    ]),
    ancestorUnitIds: ['department-it', 'service-operations'],
    pathNames: ['IT', 'Operations', 'Help Desk'],
    scopeKeys: ['org:org-arptc', 'unit:bureau-helpdesk'],
  };
  assert.throws(
    () => buildAssignmentDocument({
      assignmentId: 'head-1',
      agentId: 'uid-1',
      unit,
      assignmentType: 'HEAD',
      isPrimary: false,
      isActing: true,
      startsAt: new Date(),
      reason: 'Acting appointment',
      actorUid: 'manager-1',
    }),
    (error) => error.code === 'invalid-argument' &&
      error.message.includes('end date'),
  );
});

function withId(id, data) {
  return { id, ...data };
}

function activeUnit(id, type, name, pathUnitIds) {
  return {
    id,
    organizationId: organization.id,
    type,
    name,
    status: 'ACTIVE',
    pathUnitIds,
  };
}
